### ISSUE TITLE
Error adding an object to a collection involved in a DIRECT many to many relationship

### ISSUE THREAD

**Issue by dannief** [NakedObjectsGroup/NakedObjectsFramework#422]
Error occurs when invoking the contributed action `AssignSquad`. See below

```c#

public class TrackingContributedActions
{        
  public void AssignSquad([ContributedAction] System system, Squad squad)
  {            
      system.Squads.Add(squad);
  }
}

public partial class Squad
{
  public virtual int Id { get; set; }
 
  public virtual string Name { get; set; } = null!;
  
  public virtual ICollection<System> Systems { get; set; } = new HashSet<System>();
}

public partial class System
{   
    public virtual int Id { get; set; }

    public virtual string Name { get; set; } = null!;
    
    public virtual ICollection<Squad> Squads { get; set; } = new HashSet<Squad>();
}
```
```C#
public class AppDbContext 
{
  protected override void OnModelCreating(ModelBuilder modelBuilder)
  {
      modelBuilder.Entity<Squad>(entity =>
      {
          entity.ToTable("squad", "people");

          entity.Property(e => e.Id).HasColumnName("id");

          entity.Property(e => e.Name)
              .HasColumnType("character varying")
              .HasColumnName("name")
              .HasComment("Official name of the squad");
        });

        modelBuilder.Entity<System>(entity =>
        {
              entity.ToTable("system", "tracking");

              entity.Property(e => e.Id).HasColumnName("id");

              entity.Property(e => e.Name)
                  .HasColumnType("character varying")
                  .HasColumnName("name");

              entity.HasMany(d => d.Squads)
                  .WithMany(p => p.Systems)
                  .UsingEntity<Dictionary<string, object>>(
                      "SystemSquadAssignment",
                      l => l.HasOne<Squad>().WithMany().HasForeignKey("SquadId").HasConstraintName("system_squad_assignment_fk_squad_id"),
                      r => r.HasOne<System>().WithMany().HasForeignKey("SystemId").HasConstraintName("system_squad_assignment_fk_system_id"),
                      j =>
                      {
                          j.HasKey("SystemId", "SquadId").HasName("system_squad_assignment_pk");

                          j.ToTable("system_squad_assignment", "tracking").HasComment("Associates a system to it's assigned squads");

                          j.IndexerProperty<int>("SystemId").HasColumnName("system_id");

                          j.IndexerProperty<int>("SquadId").HasColumnName("squad_id");
                      });
          });
   }
}
```

```
Error message received from server
Message: Object reference not set to an instance of an object.
Code: InternalServerError(500)
Description: A software error has occurred on the server.
Stack Trace :
at NakedFramework.Persistor.EFCore.Util.EFCoreHelpers.GetPrimaryKey(IEntityType eType)
at NakedFramework.Persistor.EFCore.Util.EFCoreHelpers.GetKeyValues(DbContext context, Object obj)
at NakedFramework.Persistor.EFCore.Component.EFCoreObjectStore.LoadObjectIntoNakedObjectsFramework(Object domainObject, DbContext context)
at NakedFramework.Persistor.EFCore.Component.EFCoreObjectStore.<>c__DisplayClass52_0.<SetupContexts>b__2(Object _, EntityTrackedEventArgs args)
at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.OnTracked(InternalEntityEntry internalEntityEntry, Boolean fromQuery)
at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.FireStateChanged(EntityState oldState)
at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetEntityState(EntityState oldState, EntityState newState, Boolean acceptChanges, Boolean modifyProperties)
at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetEntityState(EntityState entityState, Boolean acceptChanges, Boolean modifyProperties, Nullable`1 forceStateWhenUnknownKey)
at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.NavigationFixer.FindOrCreateJoinEntry(ValueTuple`5 arguments)
at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.NavigationFixer.NavigationCollectionChanged(InternalEntityEntry entry, INavigationBase navigationBase, IEnumerable`1 added, IEnumerable`1 removed)
at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntryNotifier.NavigationCollectionChanged(InternalEntityEntry entry, INavigationBase navigationBase, IEnumerable`1 added, IEnumerable`1 removed)
at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.DetectNavigationChange(InternalEntityEntry entry, INavigationBase navigationBase)
at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.LocalDetectChanges(InternalEntityEntry entry)
at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.DetectChanges(IStateManager stateManager)
at Microsoft.EntityFrameworkCore.ChangeTracking.ChangeTracker.DetectChanges()
at NakedFramework.Persistor.EFCore.Component.EFCoreLocalContext.PreSave()
at NakedFramework.Persistor.EFCore.Component.EFCoreObjectStore.<>c.<PreSave>b__66_0(EFCoreLocalContext c)
at System.Array.ForEach[T](T[] array, Action`1 action)
at NakedFramework.Core.Util.CollectionUtils.ForEach[T](T[] toIterate, Action`1 action)
at NakedFramework.Persistor.EFCore.Component.EFCoreObjectStore.PreSave()
at NakedFramework.Persistor.EFCore.Component.EFCoreObjectStore.RecurseUntilAllChangesApplied(Int32 depth)
at NakedFramework.Persistor.EFCore.Component.EFCoreObjectStore.EndTransaction()
at NakedFramework.Core.Transaction.NestedTransaction.Commit()
at NakedFramework.Core.Component.TransactionManager.EndTransaction()
at NakedFramework.Facade.Impl.Impl.FrameworkFacade.End(Boolean success)
at NakedFramework.Rest.API.RestfulObjectsControllerBase.InitAndHandleErrors(Func`1 f)

```

**Comment by richardpawson** [comment]
As you have quite a complex mapping for the link table which I am not familiar with, can you please  try it without that mapping (i.e. so that EFC just creates the default link table) - and confirm whether you get the same problem?

**Comment by richardpawson** [comment]
We have reproduced and now fixed this issue, pending release.  The problem was specific to _direct_ many-to-many relationships, and our tests for many-to-many were all indirect (i.e. where the link table is mapped as a domain object - typically because you want additional attributes such as applicable dates stored).  Direct many-to-many relationships had been working fine with EF6, but it seems that EF Core handles them slightly differently.

**Comment by richardpawson** [comment]
We have now published a v13.0.0-beta04 version of the Server and ProgrammingModel packages that fixes this issue (we used your actual code example in the test).
